ResearchKit/ActiveTasks/ORKHeartRateCameraRecorder.m

```diff
@@ -0,0 +1,609 @@
+ the copyright holders even if such marks are included in this software.
+ 
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+ FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "ORKHeartRateCameraRecorder.h"
+@import AVFoundation;
+
+#import "ORKDataLogger.h"
+
+#import "ORKRecorder_Internal.h"
+
+#import "ORKHelpers_Internal.h"
+
+#import "ORKCodingObjects.h"
+
+const NSTimeInterval ORKHeartRateSampleRate = 1.0;
+const int ORKHeartRateFramesPerSecond = 30;
+const int ORKHeartRateMaxWindowSeconds = 30;
+const int ORKHeartRateMinWindowSeconds = 5;
+
+NSString * const ORKTimestampKey = @"timestamp";
+NSString * const ORKColorHueKey = @"hue";
+NSString * const ORKColorSaturationKey = @"saturation";
+NSString * const ORKColorBrightnessKey = @"brightness";
+NSString * const ORKColorRedKey = @"red";
+NSString * const ORKColorGreenKey = @"green";
+NSString * const ORKColorBlueKey = @"blue";
+NSString * const ORKCameraHeartRateKey = @"bpm";
+NSString * const ORKWatchHeartRateKey = @"bpm_watch";
+
+@interface ORKHeartRateCameraRecorder() <AVCaptureVideoDataOutputSampleBufferDelegate>
+
+@end
+
+@implementation ORKHeartRateCameraRecorder {
+    BOOL _started;
+    ORKDataLogger *_logger;
+#if TARGET_OS_SIMULATOR
+    NSTimer *_simulationTimer;
+#else
+    AVCaptureSession *_session;
+#endif
+    NSMutableArray *_dataPointsHue;
+    NSMutableArray *_samples;
+    NSTimeInterval _timestamp;
+}
+
+- (void)dealloc {
+    ORK_Log_Debug(@"Remove session %p", self);
+#if TARGET_OS_SIMULATOR
+    [_simulationTimer invalidate];
+    _simulationTimer = nil;
+#else
+    [_session stopRunning];
+#endif
+    [_logger finishCurrentLog];
+}
+
+#pragma mark - Data collection
+
+- (NSString *)recorderType {
+    return @"heartrate";
+}
+
+- (BOOL)isRecording {
+#if TARGET_OS_SIMULATOR
+    return (_simulationTimer != nil);
+#else
+    return _session.isRunning;
+#endif
+}
+
+- (void)start {
+    if (_started) {
+        [super start];
+        return;
+    }
+    _started = YES;
+    
+    if (!_logger) {
+        NSError *error = nil;
+        _logger = [self makeJSONDataLoggerWithError:&error];
+        if (!_logger) {
+            [self finishRecordingWithError:error];
+            return;
+        }
+    }
+    
+    _timestamp = 0;
```

Comment:
sample rate is constant in this case so the timestamp is relative to when the process was started. This is also required in order to be able to compare the heart rate that is calculated to the heart rate from the watch. In other words, the system clock isn't really relevant. That said, a better method would probably be to always calculate the timestamp based on the running total, so I will change that.